### PR TITLE
Add an Atom::from_addrefed function.

### PR DESCRIPTION
**Pull request by heycam**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16269)

### CODE REVIEW

**Comment by heycam** [comment]
Hopefully this fixes the current leak on autoland.

https://treeherder.mozilla.org/#/jobs?repo=try&revision=e8e42f28bea90ba22954239eb1aa56331fa560ba

If that try run looks good: r? @emilio or @Manishearth

**Comment by heycam** [comment]
(Feel free to r+ directly rather than delegate+ since I might not be around to land this myself, and there is no Gecko side change.)

File: components/style/gecko_string_cache/mod.rs [hunk 1]
```diff
@@ -198,6 +198,16 @@ impl Atom {
                       "Called from_static for a non-static atom!");
         atom
     }
+
+    /// Creates an atom from a dynamic atom pointer that has already had AddRef
+    /// called on it.
+    #[inline]
+    pub fn from_addrefed(ptr: *mut nsIAtom) -> Self {
```
Comment (nox) [inline components/style/gecko_string_cache/mod.rs]: This should be unsafe.

**Review by nox — CHANGES_REQUESTED** [review]

**Comment by heycam** [comment]
The try push above was on top of another broken patch which had its own leaks.  But according to @upsuper it it looks like the two nsStringBuffers that this patch was meant to fix are indeed fixed.

**Pull request merged**